be/src/storage/tablet_reader.cpp

```diff
@@ -334,10 +334,15 @@ Status TabletReader::_init_delete_predicates(const TabletReaderParams& params, D
     PredicateParser pred_parser(_tablet->tablet_schema());
 
     std::shared_lock header_lock(_tablet->get_header_lock());
-    for (const DeletePredicatePB& pred_pb : _tablet->delete_predicates()) {
-        if (pred_pb.version() > _delete_predicates_version.second) {
+    // here we can not use DeletePredicatePB from  _tablet->delete_predicates() because
+    // _rowsets maybe stale rowset, and stale rowset's delete predicates may be removed
+    // from _tablet->delete_predicates() after compation
+    for (const RowsetSharedPtr& rowset : _rowsets) {
+        const RowsetMetaSharedPtr& rowset_meta = rowset->rowset_meta();
+        if (!rowset_meta->has_delete_predicate()) {
             continue;
         }
+        const DeletePredicatePB& pred_pb = rowset_meta->delete_predicate();
```

Comment:
Why don't you check `if (pred_pb.version() > _delete_predicates_version.second) {` any more?

Reply:
Because there delete_predicates are from rowsets that captured(capture_consistent_rowsets)  using _delete_predicates_version (_delete_predicates_version = _version = version)
It can make sure that pred_pb.version() <= _delete_predicates_version.second